### ISSUE TITLE
cmd/fmt: invalid CUE emitted when all fields are commented inside struct

### ISSUE THREAD

**Issue by jpluscplusm** [cue-lang/cue#2423]
### What version of CUE are you using (`cue version`)?

<pre>
$ cue version
cue version v0.6.0-alpha.1

go version go1.20.3
      -buildmode exe
       -compiler gc
       -trimpath true
     CGO_ENABLED 0
          GOARCH amd64
            GOOS linux
         GOAMD64 v1
</pre>

### Does this issue reproduce with the latest stable release?

Yes: 0.5.0 also fails.

### What did you do?

```testscript
exec cue vet .:p
exec cue fmt .:p
exec cue vet .:p

-- file.cue --
package p

#X: {a?: int}

foo: #X & {
  // a: 5
}
```

```shell
testscript fmt.txtar
> exec cue vet .:p
> exec cue fmt .:p
> exec cue vet .:p
[stderr]
expected operand, found '&':
    ./file.cue:7:2
[exit status 1]
FAIL: /tmp/testscript743196369/fmt.txtar/script.txtar:3: unexpected command failure
error running fmt.txtar in /tmp/testscript743196369/fmt.txtar
```

### What did you expect to see?

I expected `cue fmt` to emit valid CUE.

### What did you see instead?

The 2nd `cue vet` fails, indicating that `cue fmt` emitted invalid CUE.

A quick pass over the Issues in milestone `fmt-redesign` didn't reveal any already capturing this problem.

**Comment by myitcv** [comment]
Potentially a duplicate. Will double check when we come to triage.

**Comment by jpluscplusm** [comment]
@myitcv You might be thinking of #2274, which I also opened. I opened this issue separately, as the triggering setup and invalid CUE emitted both look different to my eyes.

I don't have any insight into the underlying similarity, so of course please close as a dupe or leave as a tracker however best suits the project :-)

**Comment by mvdan** [comment]
Yeah, I also seemed to recall a similar issue, and I think that was the one :)

**Comment by NoamTD** [comment]
I simplified the example to:

```cue
example: {} & {
    // This is a comment
}
```

The output of `cue fmt` in this case is:
```cue
example: {}
	// my comment
	& {
	}
```

I narrowed down the issue to the cue parser, which incorrectly(?) associates the comment with the `example` node, instead of the rhs struct

**Comment by mvdan** [comment]
Note that this issue is closely related to https://github.com/cue-lang/cue/issues/2567 and https://github.com/cue-lang/cue/issues/2274; it's likely that a fix in `cue/format` would fix multiple of these reproducers. cc @uhthomas in case you are looking into any of these already.

**Comment by rudifa** [comment]
Looking at @NoamTD's example 
```
cue % cat testdata/2423-2.cue                                      [issues-fmt-comments L|…1]
example: {} & {
    // This is a comment
}
```
... with fmt.go instrumented to print `DebugStr(f)` after the decoder shows that the comment gets associated with the first `{}` and then the encoder misplaces the comment
```
cue % cat testdata/2423-2.cue | go run . fmt                        [issues-fmt-comments L|…1]
example: {}
	// This is a comment
	& {
	}
```
```
DebugStr: example: <[d1// This is a comment] {}&{}>
```
... and generates invalid `cue`.

Re `print `DebugStr(f)`` see my comment in #1447.

**Comment by rudifa** [comment]
Building on the @NoamTD's example, this reproducer demonstrates a few facts about the #2423 bug:
1. in the context of `&` operation, a comment in an empty list (`ex1`) is treated similarly to a comment in an empty struct (`ex0`): both are misplaced, producing invalid cue.
2. in a simpler context, as values in a top level field, both the empty struct (`ex2`) and the empty list (`ex3`) handle the comment correctly.
3. DebugStr shows that the offending comments are promoted to Doc comments at the binary `&` operation level, instead of remaining inside the struct or list. 
```
 issue #2423

exec cue vet 2423.cue
exec cue fmt 2423.cue
cmp 2423.cue 2423.got.cue   # shows the misplaced comment in ex0 and ex1
! exec cue vet 2423.cue     # the result of cue fmt fails vet
cmp stderr 2423.stderr

-- 2423.cue --
ex0: {
	v: 10 // ex0: comment after a value
} & {
	// ex0: comment in an empty struct
}

ex1: [
	11, // ex1: comment after a value
] & [
	// ex1: comment in an empty list
	...
]

ex2: {
	// ex2: comment in an empty struct
}

ex3: [
	// ex3: comment in an empty list
]
-- 2423.got.cue --
ex0: {
	v: 10 // ex0: comment after a value
}
	// ex0: comment in an empty struct
	& {
	}

ex1: [
	11, // ex1: comment after a value
]
	// ex1: comment in an empty list
	& [
		...,
	]

ex2: {
	// ex2: comment in an empty struct
}

ex3: [
	// ex3: comment in an empty list
]
-- 2423.stderr --
expected operand, found '&':
    ./2423.cue:5:2
-- end --
```

DebugStr (with newlines inserted for clarity):
```
 <[d0// 2423/2423.comment.in.empty.struct.BAD.cue] 
ex0: <[d1// ex0: comment in an empty struct] {<[l5// ex0: comment after a value] v: 10>}&{}>>, 
ex1: <[d1// ex1: comment in an empty list] [<[l2// ex1: comment after a value] 11>]&[...]>, 
ex2: <[d1// ex2: comment in an empty struct] {}>, 
ex3: <[d1// ex3: comment in an empty list] []>
```

**Comment by cuematthew** [comment]
Fixed by https://cue.gerrithub.io/c/cue-lang/cue/+/1235396

My testing shows that none of the examples in this issue lose their comments.
